src/abstract_domains.rs

```diff
@@ -32,9 +112,93 @@ pub enum ExpressionDomain {
+        match self {
+            ExpressionDomain::Bottom => true,
+            _ => false,
+        }
+    }
+
+    /// True if all possible concrete values are elements of the set corresponding to this domain.
+    fn is_top(&self) -> bool {
+        match self {
+            ExpressionDomain::Top => true,
+            _ => false,
+        }
+    }
+
+    /// Returns a domain whose corresponding set of concrete values include all of the values
+    /// corresponding to self and other.
+    /// In a context where the join condition is known to be true, the result can be refined to be
+    /// just self, correspondingly if it is known to be false, the result can be refined to be just other.
+    fn join(&self, other: &ExpressionDomain, join_condition: &AbstractDomains) -> ExpressionDomain {
+        ExpressionDomain::ConditionalExpression {
+            condition: box join_condition.clone(),
+            if_true: box self.clone(),
+            if_false: box other.clone(),
+        }
+    }
+
+    /// True if all of the concrete values that correspond to self also correspond to other.
+    /// Note: !x.subset(y) does not imply y.subset(x).
+    fn subset(&self, other: &ExpressionDomain) -> bool {
+        match (self, other) {
+            // The empty set is a subset of every other set
+            (ExpressionDomain::Bottom, _) => true,
+            // A non empty set if not a subset of the empty set
+            (_, ExpressionDomain::Bottom) => false,
+            // Every set is a subset of the universal set
+            (_, ExpressionDomain::Top) => true,
+            // The universal set is not a subset of any set other than the universal set
+            (ExpressionDomain::Top, _) => false,
+            (
+                ExpressionDomain::ConditionalExpression {
+                    if_true, if_false, ..
+                },
+                _,
+            ) => {
+                // This is a conservative answer. False does not imply other.subset(self).
+                if_true.subset(other) && if_false.subset(other)
+            }
+            (
+                _,
+                ExpressionDomain::ConditionalExpression {
+                    if_true, if_false, ..
+                },
+            ) => {
+                // This is a conservative answer. False does not imply other.subset(self).
+                self.subset(if_true) || self.subset(if_false)
+            }
+            // {x} subset {y} iff x = y
+            (
+                ExpressionDomain::CompileTimeConstant(cv1),
+                ExpressionDomain::CompileTimeConstant(cv2),
+            ) => cv1 == cv2,
+        }
+    }
+
+    /// Returns a domain whose corresponding set of concrete values include all of the values
+    /// corresponding to self and other.The set of values may be less precise (more inclusive) than
+    /// the set returned by join. The chief requirement is that a small number of widen calls
+    /// deterministically lead to Top.
```

Comment:
To be pedantic: widening needs to converge to a fixed point that's a sound approximation, but not necessarily to Top. Going to Top is certainly a safe choice for implementing this more permissive specification, though :).

Reply:
Yes, and I like safe. :-)